File: src/modules/fix.rs
use crate::args::FixesArgs;
use colored::Colorize;
use std::fs::File;
use std::io::prelude::*;
use std::process::Command;
use std::{env, fs};

pub fn fix(args: &FixesArgs) {
    if args.discord {
        println!("{}", "Fixing Discord Flatpak RPC...".bold().yellow());

        fix_rpc("com.discordapp.Discord", "discord", "Discord");

        println!("{}", "Done!".bold().green());
    }

    if args.gmod {
        println!("{}", "Fixing gmod...".bold().yellow());

        Command::new("mkdir")
            .args(["-p", "/tmp/patch-gmod"])
            .status()
            .expect("Weren't able to create the tmp dir");

        Command::new("sh")
            .args(["-c", "wget $(curl -s https://api.github.com/repos/solsticegamestudios/GModCEFCodecFix/releases/latest | jq -r '.assets[] | select(.name | test(\"GModCEFCodecFix-Linux\")) | .browser_download_url') -P /tmp/patch-gmod"])
            .status()
            .expect("Wasn't able to fix GMod");

        Command::new("chmod")
            .args(["+x", "/tmp/patch-gmod/GModCEFCodecFix-Linux"])
            .status()
            .expect("Wasn't able to chmod");

        Command::new("/tmp/patch-gmod/GModCEFCodecFix-Linux")
            .status()
            .expect("Not able to run script");

        Command::new("rm")
            .args(["-rf", "/tmp/patch-gmod"])
            .status()
            .expect("Not able to remove /tmp/patch-gmod");

        println!("{}", "Done!".bold().green());
    }

    if args.proton_hang {
        println!("{}", "Fixing proton hang...".bold().yellow());

        let protoncore = [
            "pv-bwrap",
            "pressure-vessel",
            "reaper",
            "explorer.exe",
            "rpcss.exe",
            "plugplay.exe",
            "services.exe",
            "svchost.exe",
            "winedevice.exe",
            "winedevice.exe",
            "wineserver",
        ];

        for x in protoncore {
            Command::new("killall")
                .args(["-9", x])
                .status()
                .expect("Was unable to kill");
        }

        println!("{}", "Done!".bold().green());
    }

    if args.vesktop {
        println!("{}", "Fixing Vesktop Flatpak RPC...".bold().yellow());

        fix_rpc("dev.vencord.Vesktop", "vesktop", "Vesktop");

        println!("{}", "Done!".bold().green());
    }
}

fn fix_rpc(long_client: &str, short_client: &str, pretty_name: &str) {
    let user = env::var("USER").unwrap();
    let tmp_dir = format!("/home/{}/.config/user-tmpfiles.d", &user);

    match fs::create_dir(&tmp_dir) {
        Ok(_) => {
            println!("{}", "Directory created...".italic().yellow());
        }
        Err(_) => {
            println!(
                "{}",
                "Directory already exists, continuing...".italic().yellow()
            );
        }
    }

    let mut tmp_file =
        File::create(format!("{}/discord-rpc.conf", tmp_dir)).expect("Wasn't able to create file");
    tmp_file
        .write_all(
            format!(
                "L %t/discord-ipc-0 - - - - app/{}/discord-ipc-0",
                long_client
            )
            .as_bytes(),
        )
        .expect("Wasn't able to write to file");

    Command::new("systemctl")
        .args([
            "enable",
            "--user",
            "--now",
            "systemd-tmpfiles-setup.service",
        ])
        .status()
        .expect("Unable to start the service");

    Command::new("ln")
        .args([
            "-s",
            format!(
                "/run/user/1000/.flatpak/{}/xdg-run/discord-ipc-0",
                long_client
            )
            .as_str(),
            "/run/user/1000/discord-ipc-0",
        ])
        .status()
        .expect("Unable to link");

    println!("{}", "Need to use sudo...".bold().blue());

    Command::new("sudo")
        .args(["flatpak", "override", "--filesystem=xdg-run/discord-ipc-*"])
        .status()
        .expect("Unable to override discord-ipc");

    Command::new("sudo")
        .args([
            "flatpak",
            "override",
            format!("--filesystem=xdg-run/.flatpak/{}:create", long_client).as_str(),
        ])
        .status()
        .expect(format!("Unable to override {}", &pretty_name).as_str());

    println!("{}", "Adding fix to autostart...".italic().blue());

    Command::new("mkdir")
        .args(["-pv", format!("/home/{}/.config/autostart", user).as_str()])
        .status()
        .expect("Unable to make autostart dir");

    let mut autostart_file = File::create(format!(
        "/home/{}/.config/autostart/{}-rpc.desktop",
        user, short_client
    ))
    .expect("Wasn't able to create file");
    autostart_file
      .write_all(format!("[Desktop Entry]\nCategories=Utility;\nComment=Symlink for {pretty_name} RPC\nIcon={short_client}\nName={pretty_name}RPC\nStartupNotify=true\nTerminal=false\nType=Application\nExec=ln -s /run/user/1000/.flatpak/{long_client}/xdg-run/discord-ipc-0 /run/user/1000/discord-ipc-0").as_bytes())
      .expect("Wasn't able to write to file");
}

File: cmd/fix.go
package cmd

import (
	"github.com/spf13/cobra"
)

//! DON'T ADD ANY SUBCOMMANDS AND STUFF HERE, CREATE A FILE IN fixes/ FOR IT.
//! CHECK THE CURRENT FILES IN THERE FOR AN EXAMPLE
//!
//! IF YOU WANT TO MAKE ONE SPECIFIC FOR AN IMAGE, IT GOES IN THE IMAGE FOLDER (horizon/nova/umbra) LIKE NORMAL.
//! JUST BE SURE TO INIT WITH cmd.FixCmd.AddCommand()

var (
	FixCmd = &cobra.Command{
		Use:   "fix",
		Short: "Different fixes for various things",
		Args:  cobra.ExactArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
		},
	}

	// ATM THE FIXES FOR THESE DOESN'T REALLY WORK, WILL BE IN HERE FOR THE TIME BEING

	// fixDiscord = &cobra.Command{
	// 	Use:   "discord",
	// 	Short: "Fix Discord flatpak RPC",
	// 	Run: func(cmd *cobra.Command, args []string) {
	// 		fixRpc("com.discordapp.Discord", "discord", "Discord")
	// 	},
	// }

	// fixVesktop = &cobra.Command{
	// 	Use:   "vesktop",
	// 	Short: "Fix Vesktop flatpak RPC",
	// 	Run: func(cmd *cobra.Command, args []string) {
	// 		fixRpc("dev.vencord.Vesktop", "vesktop", "Vesktop")
	// 	},
	// }
)

func init() {
	RootCmd.AddCommand(FixCmd)
	// fixCmd.AddCommand(fixDiscord)
	// fixCmd.AddCommand(fixVesktop)
}

// func fixRpc(longClient string, shortClient string, prettyname string) {
// 	configDir, err := os.UserConfigDir()
// 	if err != nil {
// 		utils.Error("Unable to get home directory")
// 		os.Exit(1)
// 	}

// 	tmpDir := path.Join(configDir, "user-tmpfiles.d")
// 	if err := os.MkdirAll(tmpDir, os.ModePerm); err != nil {
// 		utils.Error("Failed to create: " + tmpDir)
// 		os.Exit(1)
// 	}

// 	if shortClient == "vesktop" {
// 		if err := utils.WriteFile(path.Join(tmpDir, "discord-rpc.conf"), "L %t/discord-ipc-0 - - - - .flatpak/dev.vencord.Vesktop/xdg-run/discord-ipc-0"); err != nil {
// 			utils.Error("Was unable to create and write to file")
// 			os.Exit(1)
// 		}
// 	} else if shortClient == "discord" {
// 		if err := utils.WriteFile(path.Join(tmpDir, "discord-rpc.conf"), "L %t/discord-ipc-0 - - - - app/com.discordapp.Discord/discord-ipc-0"); err != nil {
// 			utils.Error("Was unable to create and write to file")
// 			os.Exit(1)
// 		}
// 	}

// 	cmdToRun := exec.Command("sudo", "systemctl", "enable", "--user", "--now", "systemd-tmpfiles-setup.service")
// 	if err := cmdToRun.Run(); err != nil {
// 		utils.Error("Was unable to enable systemd-tmpfiles-setup.service")
// 		os.Exit(1)
// 	}

// 	if utils.PathExists("/run/user/1000/discord-ipc-0") {
// 		os.Remove("/run/user/1000/discord-ipc-0")
// 	}

// 	if err := os.Symlink("/run/user/1000/.flatpak/"+longClient+"/xdg-run/discord-ipc-0", "/run/user/1000/discord-ipc-0"); err != nil {
// 		utils.Error("Was unable to create symlink")
// 		os.Exit(1)
// 	}

// 	utils.Notice("Will need to use sudo...")

// 	cmdToRun = exec.Command("sudo", "flatpak", "override", "--filesystem=xdg-run/discord-ipc-*")
// 	if err := cmdToRun.Run(); err != nil {
// 		utils.Error("Was unable to override discord-ipc")
// 		os.Exit(1)
// 	}

// 	cmdToRun = exec.Command("sudo", "flatpak", "override", "--filesystem=xdg-run/.flatpak/"+longClient+":create")
// 	if err := cmdToRun.Run(); err != nil {
// 		utils.Error("Was unable to override " + prettyname)
// 		os.Exit(1)
// 	}

// 	utils.Info("Adding fix to autostart...")

// 	autostartDir := path.Join(configDir, "autostart")
// 	if err := os.MkdirAll(autostartDir, os.ModePerm); err != nil {
// 		utils.Error("Unable to make: " + autostartDir)
// 		os.Exit(1)
// 	}

// 	if err := utils.WriteFile(path.Join(autostartDir, shortClient+"-rpc.desktop"), "[Desktop Entry]\nCategories=Utility;\nComment=Symlink for "+prettyname+" RPC\nIcon="+shortClient+"\nName="+prettyname+"RPC\nStartupNotify=true\nTerminal=false\nType=Application\nExec=ln -s /run/user/1000/.flatpak/"+longClient+"/xdg-run/discord-ipc-0 /run/user/1000/discord-ipc-0"); err != nil {
// 		utils.Error("Was unable to create and write to file")
// 		os.Exit(1)
// 	}

// 	utils.Done("Fixed " + prettyname + " RPC")
// }
